### PR TITLE
[#3336] do not add not-found tool paths

### PR DESCRIPTION
**Pull request by mwichmann**
Resolve Issue #3336 

When tool modules initialize, they check paths to decide if the underlying tool is actually present.  The current checker adds any "default paths" the caller may have supplied (locations like mingw, cygwin, chocolatey install locations, etc.); if there is match from this list, any previous default paths are also kept. To avoid keeping these non-matching paths, restore the original PATH; the caller is responsible for adding to PATH if necessary. Docstring now says so.

Note lex and yacc tool modules expected the path-modifying behavior that's being gotten rid of - so they preseve the path first and restore it after.  That has been rolled back out.

The `swig.py` change is not functional, it just aligns the usage with what is introduced elsewhere.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [X] I have updated the appropriate documentation

### CODE REVIEW

**Comment by mwichmann** [comment]
The one Appveyor fail is from a file close issue for which a patch exists elsewhere, not sure why it's intermittent.

```
Failed the following test:
	test\Interactive\configure.py
```

```
193/300 (64.33%) C:\\Python37\\python.exe test\Interactive\configure.py
FAILED test of C:\projects\scons\src\script\scons.py
	at line 615 of C:\projects\scons\testing\framework\TestCommon.py (_complete)
	from line 675 of C:\projects\scons\testing\framework\TestCommon.py (finish)
	from line 119 of test\Interactive\configure.py
match_re: mismatch at line 2:
  search re='^scons>>> scons: `foo.obj' is up to date.$'
  line='scons>>> C:\Python37\python.exe mycc.py foo.obj foo.cpp'
STDOUT =========================================================================
1,2c1,3
```

File: src/engine/SCons/Tool/yacc.py [hunk 1]
```diff
@@ -100,11 +100,15 @@ def yyEmitter(target, source, env):
...
+    Only called in the Windows case, so the default_path
+    can be Windows-specific
+
+    :param env: current construction environment
+    :param append_paths: if set, add the path to the tool to PATH
+    :return: path to yacc tool, if found
     """
-    # save existing path to reset if we don't want to append any paths
-    envPath = env['ENV']['PATH']
     bins = ['bison', 'yacc', 'win_bison']
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/yacc.py]: Can you move this list to the top, name win_bins and have a non-win bins and then use that everywhere instead of having several lists scattered throughout the module?

Comment (mwichmann) [reply src/engine/SCons/Tool/yacc.py]: no problem, I've got that ready to push... but while I'm at it, the block in `yacc.py` from lines 145-151 seems like it is redundant now - like the "old version" was still left around when the `get_yacc_path` function was added. Should this be dropped?

Comment (bdbaddog) [reply src/engine/SCons/Tool/yacc.py]: Sure.

**Comment by mwichmann** [comment]
fwiw, the build fail from appveyor is surpious - something happend during the chocolatey setup, it wasn't from the tests themselves.

**Comment by mwichmann** [comment]
So if this test is "wrong", I don't have any other ideas how to test.  I don't get how to check a filesystem glob to match jdks without having something in the filesystem to glob against.  Maybe just give up on the test?

**Comment by bdbaddog** [comment]
Unit test's don't touch the filesystem, or shell out and run commands.
I think you should be able to monkey patch glob.. lemme fork your branch and give it a try.
Since the change is really about finding specific directory patterns.
That should be unit testable.. (I think..)

File: src/engine/SCons/Tool/__init__.py [hunk 2]
```diff
@@ -394,7 +396,8 @@ def _call_env_subst(env, string, *args, **kw):
 
 
 class _ShLibInfoSupport(object):
-    def get_libtype(self):
+    @property
+    def libtype(self):
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/__init__.py:400]: Is this related to the bug?
